### PR TITLE
Fix plan rejection and error handling in deploy workflow.

### PR DESCRIPTION
**Pull request by nishkrishnan**
Went through all our error handling and error logging and this is what I came up with. Feel free to ask questions if it's not clear.

- [x] Finish writing tests

### CODE REVIEW

File: server/neptune/workflows/internal/deploy/terraform/runner.go [hunk 1]
```diff
@@ -31,6 +33,11 @@ func (r *WorkflowRunner) Run(ctx workflow.Context, deploymentInfo DeploymentInfo
 	id := deploymentInfo.ID
 	ctx = workflow.WithChildOptions(ctx, workflow.ChildWorkflowOptions{
 		WorkflowID: id.String(),
+		// we shouldn't ever percolate failures up unless they are user errors (aka. Terraform specific)
+		// retrying indefinitely allows us to fix whatever issue that comes about without involving the user to redeploy
+		RetryPolicy: &temporal.RetryPolicy{
+			NonRetryableErrorTypes: []string{"TerraformClientError", "PlanRejectedError"},
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/deploy/terraform/runner.go:39]: I haven't looked to closely at the SDK code, but by default Workflow Executions don't retry: https://docs.temporal.io/application-development/features#workflow-retries, do you need to specify the other RetryPolicy parameters to create the full policy?

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 2]
```diff
@@ -45,22 +45,13 @@ const (
...
-	HeartBeatTimeout       = 1 * time.Minute
-	TerraformClientError   = "TerraformClientError"
+	PlanReviewSignalName     = "planreview"
+	ScheduleToCloseTimeout   = 30 * time.Minute
+	HeartBeatTimeout         = 1 * time.Minute
+	TerraformClientErrorName = "TerraformClientError"
 )
 
 func Workflow(ctx workflow.Context, request Request) error {
-	options := workflow.ActivityOptions{
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/terraform/workflow.go:55]: Without this, I believe we go back to the defaults for the first activity of the workflow.

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 3]
```diff
@@ -236,14 +232,37 @@ func (r *Runner) Run(ctx workflow.Context) error {
 	}()
 
 	planResponse, err := r.Plan(ctx, root, response.ServerURL)
-
 	if err != nil {
-		return errors.Wrap(err, "running plan job")
+		return toExternalError(err, "running plan job")
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/terraform/workflow.go:236]: We only get a plan rejected error within the Apply fxn right?

File: server/neptune/workflows/internal/deploy/terraform/runner.go [hunk 4]
```diff
@@ -50,15 +57,14 @@ func (r *WorkflowRunner) Run(ctx workflow.Context, deploymentInfo DeploymentInfo
...
-		return errors.Wrap(err, "getting child workflow execution")
-	}
-
-	selector := workflow.NewSelector(ctx)
+	selector := workflow.NewNamedSelector(ctx, "TerraformChildWorkflow")
 
-	// our child workflow will signal us when there is a state change which we will
-	// handle accordingly
+	// our child workflow will signal us when there is a state change which we will handle accordingly.
+	// if for some reason the workflow is orphaned or we are retrying it independently, we have no way
```
Comment (samrabelachew) [inline server/neptune/workflows/internal/deploy/terraform/runner.go:63]: Can you explain an example of how the child wf gets orphaned? I imagine we would want to fail any running terraform workflows if the parent deploy fails.

Just wanted to highlight that there is a field in the ChildWorkflowOptions that also gives us the ability to cancel a workflow if the parent workflow fails for some reason: https://github.com/temporalio/api-go/blob/v1.12.0/enums/v1/workflow.pb.go#L87

File: server/neptune/workflows/internal/terraform/workflow.go [hunk 5]
```diff
@@ -236,14 +232,37 @@ func (r *Runner) Run(ctx workflow.Context) error {
...
+		return toExternalError(err, "running apply job")
 	}
 
 	return nil
 }
+
+// toExternalError allows callers of the workflow to handle specific error types
+// in whatever fashion.
+// we use errors.As here to ensure that we're accounting for wrapped errors
+func toExternalError(err error, msg string) error {
```
Comment (Aayyush) [inline server/neptune/workflows/internal/terraform/workflow.go:249]: nit: thoughts on using switch statement to check for the error type?

Comment (nishkrishnan) [reply server/neptune/workflows/internal/terraform/workflow.go:265]: This has changed btw

File: server/neptune/workflows/internal/deploy/terraform/runner.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nishkrishnan) [reply server/neptune/workflows/internal/deploy/terraform/runner.go:62]: I wasn't talking about whether the parent workflow would terminate.  I was talking about when the parent workflow moves to the next deploy but the terraform workflow is orphaned.

File: server/neptune/workflows/internal/terraform/workflow.go (hunk 2 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/terraform/workflow.go:55]: Fixed this in the latest implementation.

File: server/neptune/workflows/internal/deploy/terraform/runner.go (hunk 4 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/deploy/terraform/runner.go:62]: let's say we need to kill it manually, and then we restart using the commandline, it'd be orphaned and have no where to report its state.

File: server/neptune/workflows/internal/terraform/workflow.go (hunk 5 above)
Comment (nishkrishnan) [reply server/neptune/workflows/internal/terraform/workflow.go:266]: Also not sure how i would even use a switch statement for this.

**Review by samrabelachew — APPROVED** [review]

File: server/neptune/workflows/internal/deploy/terraform/runner.go [hunk 6]
```diff
@@ -78,8 +91,17 @@ func (r *WorkflowRunner) awaitWorkflow(ctx workflow.Context, future workflow.Chi
 		}
 	}
 
-	if err != nil {
-		return errors.Wrap(err, "executing terraform workflow")
+	// if we have an app error we should attempt to unwrap it's details into our own
+	// application error and act accordingly
+	var appErr *temporal.ApplicationError
+	if errors.As(err, &appErr) {
```
Comment (Aayyush) [inline server/neptune/workflows/internal/deploy/terraform/runner.go:97]: do we also need to handle `UpdateJobError` here?

Comment (nishkrishnan) [reply server/neptune/workflows/internal/deploy/terraform/runner.go:97]: no its a retryable error type.

**Pull request merged**